cmd/flux/cmd/test.go

```diff
@@ -0,0 +1,151 @@
+package cmd
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/influxdata/flux"
+	"github.com/influxdata/flux/ast"
+	"github.com/influxdata/flux/ast/edit"
+	"github.com/influxdata/flux/codes"
+	"github.com/influxdata/flux/execute/executetest"
+	"github.com/influxdata/flux/execute/table"
+	"github.com/influxdata/flux/fluxinit"
+	"github.com/influxdata/flux/internal/errors"
+	"github.com/influxdata/flux/lang"
+	"github.com/influxdata/flux/memory"
+	"github.com/influxdata/flux/parser"
+	"github.com/influxdata/flux/runtime"
+	"github.com/spf13/cobra"
+)
+
+type Test struct {
+	ast *ast.Package
+	Err error
+}
+
+func NewTest(ast *ast.Package) Test {
+	return Test{
+		ast: ast,
+	}
+}
+
+func (t *Test) Name() string {
+	return t.ast.Files[0].Name
+}
+
+func (t *Test) Run() {
+	jsonAST, Err := json.Marshal(t.ast)
+	if Err != nil {
+		t.Err = Err
+		return
+	}
+	c := lang.ASTCompiler{AST: jsonAST}
+
+	ctx := executetest.NewTestExecuteDependencies().Inject(context.Background())
+	program, Err := c.Compile(ctx, runtime.Default)
+	if Err != nil {
+		t.Err = errors.Wrap(Err, codes.Invalid, "failed to compile")
+	}
+
+	alloc := &memory.Allocator{}
+	result, Err := program.Start(ctx, alloc)
+	if Err != nil {
+		// XXX: rockstar (8 Dec 2020) - Not all tests should return streaming data.
+		if !strings.Contains(Err.Error(), "this Flux script returns no streaming data") {
+			t.Err = errors.Wrap(Err, codes.Inherit, "error while executing program")
+			return
+		}
+	}
+	defer result.Done()
+
+	for res := range result.Results() {
```

Comment:
It might be easier to use the `ResultIterator`.

```go
qry, err := program.Start(...)
if err != nil {
    // error handling code
}

results := flux.NewResultIteratorFromQuery(qry)
defer results.Release()

for results.More() {
    res := results.Next()
    // handle the result
}
t.Err = results.Err()
```

This is the standard pattern we use in other places so I think it would be good to follow it here too.

Reply:
Ah! I _knew_ there had to be a pattern for doing this. I just couldn't find it anywhere.